client/src/components/Sidebar/common/Button.tsx

```diff
@@ -0,0 +1,70 @@
+/* eslint-disable react/prop-types */
+
+import React from 'react';
+
+const handleClick = (
+  href: string,
+  onClick: ((e: React.MouseEvent) => void) | undefined,
+  preventDefault: boolean,
+  navigate: (url: string) => Promise<void>,
+  e: React.MouseEvent
+) => {
+  if (preventDefault && href === '#' || navigate) {
+    e.preventDefault();
+    e.stopPropagation();
+  }
+
+  if (onClick) {
+    onClick(e);
+  } else if (navigate) {
+    navigate(href);
+  }
+};
+
+interface ButtonProps {
+  className?: string;
+  accessibleLabel?: string;
+  href?: string;
+  target?: string;
+  preventDefault?: boolean;
+  onClick?(e: React.MouseEvent): void;
+  dialogTrigger?: boolean;
+  navigate?(url: string): Promise<void>;
+}
+
+/**
+ * A reusable button. Uses a <a> tag underneath.
+ */
+const Button: React.FunctionComponent<ButtonProps> = ({
+  className = '',
+  children,
+  accessibleLabel,
+  href = '#',
+  target,
+  preventDefault = true,
+  onClick,
+  dialogTrigger,
+  navigate,
+}) => {
+  const hasText = children !== null;
```

Comment:
`children` in react is an array, and can be `[]`, it means that: `[] !== null // true`  
To check the presence of children, you could do the following 
```suggestion
  const hasText = children?.length > 0
  ```
`length` won't be called in children is not defined, this expression will return true only if `children` is not undefined AND it has at least a child 
```

Reply:
Unsure if we can use this syntax in Wagtail yet, will try anyway!

Reply:
> [tsl] ERROR in /home/karl/projects/wagtaildemo-sidebar-rewrite/wagtail/client/src/components/Sidebar/common/Button.tsx(49,29)
      TS2339: Property 'length' does not exist on type 'boolean | ReactChild | ReactFragment | ReactPortal'.
  Property 'length' does not exist on type 'number'.

Reply:
I think we probably still need to cope with the empty array case. You could add a type guard before the length check?

Reply:
Turns out React has a helper for finding the length, so replaced this with ``const hasText = React.Children.count(children) > 0;``